### DISCUSSION TITLE
Automatic indentation of tree-sitter doesn't work well.

### DISCUSSION THREAD

**Discussion by DrScKAWAMOTO** [emacs-tree-sitter/elisp-tree-sitter#282] [Q&A] [answered]
Dear everyone.

Automatic indentation of tree-sitter doesn't work well for me.
I tried it in two environments, and neither worked properly.

- homebrew cask emacs 29.3 on MacBookAir M1 2020 Sonoma 14.4.1
- emacs 29.1 on FreeBSD 13.2-RELEASE-p3

I have elpa package `~/.emacs.d/elpa/treesit-auto-20240511.1425` .
I have the following language bundles in `~/.emacs.d/tree-sitter` on MacBookAir,

```example
-rwxr-xr-x 1 XXXXX staff 664K  5 23 17:41 libtree-sitter-c.dylib
-rwxr-xr-x 1 XXXXX staff 3.8M  5 23 17:41 libtree-sitter-cpp.dylib
-rwxr-xr-x 1 XXXXX staff 648K  5 23 17:44 libtree-sitter-python.dylib
```

and on FreeBSD,

```example
-rwxr-xr-x 1 XXXXX XXXXX 618K May 23 23:12 libtree-sitter-c.so
-rwxr-xr-x 1 XXXXX XXXXX 3.7M May 23 23:12 libtree-sitter-cpp.so
```

I have a simple `~/.emacs.d/init.el` the follows,

```elisp
(use-package treesit
  :config
  (setq treesit-font-lock-level 4))

(use-package treesit-auto
  :ensure t
  :config
  (setq treesit-auto-install t)
  (global-treesit-auto-mode))
```

The steps to reproduce are as follows.

1. Start emacs.
2. Open the C source code file and write the following content.

```c
void test(void)
{
  do🔳
```

3. Place the cursor after `do` and press Return.

Expected behaviour is automatic indentation. I want to move the cursor to the 5th column of the next line.
However, the observed behavior is NOT indentation. The cursor moves to the 1st column of the next line.
For reference, the column position of `do` was moved correctly the 3rd column with automatic indentation.

Thanx.

**Comment by jcs090218** [comment, marked as answer]
Treesit.el and tree-sitter.el are two different implementations. Suppose you are looking to report issues to treesit.el, you should directly report to emacs-devel.

**Reply by DrScKAWAMOTO** [reply]
Oh my goodness!
Thanks for your advice.
I would like to post it to emacs-devel using the command M-x report-emacs-bug.

This raises some questions(using the c-mode for example).

1. Is it correct that emacs29+ means version 29 or higher?
2. Is treesit.el designed based on c-ts-mode, and is tree-sitter.el designed base on minor mode of c-mode?
3. Do I have to use treesit.el? I am emacs29 user.

I like c-mode auto indent implementation.
Having two major modes, c-mode and c-ts-mode, confuses me.
So I would like to use tree-sitter.el rather than treesit.el.

Thanx.

**Reply by sogaiu** [reply]
It seems like some of your questions are related to the following text from the README:

> For Emacs 29+, please use [the built-in integration](https://lists.gnu.org/archive/html/emacs-devel/2022-11/msg01443.html) instead of this package.

Assuming so...

> Is it correct that emacs29+ means version 29 or higher?

I think so.

> Do I have to use treesit.el? I am emacs29 user.

Depending on how your emacs was compiled, it may or may not have this "built-in integration".

If your emacs doesn't come with the built-in integration (as mentioned in the quote from the README), then AFAIU you won't be able to use `treesit.el`.  In such a case, if you still want to use capabilities of tree-sitter (the c library), what is here at this repository (i.e. elisp-tree-sitter) may be an option.

According to [this document](https://github.com/emacs-mirror/emacs/blob/master/admin/notes/tree-sitter/starter-guide), it looks like you can check if your build of emacs was compiled with the aforementioned built-in tree-sitter integration by evaluating `(treesit-available-p)` -- it's possible you may need to evaluate `(require 'treesit)` first, but I don't have a convenient way to verify this ATM.

I think there may be other reasons why some folks are using elisp-tree-sitter (what's provided at this repository).  It has been around a lot longer and there may be code that depends on it, so that might be one reason, but may be there are other reasons too.

> Is treesit.el designed based on c-ts-mode, and is tree-sitter.el designed base on minor mode of c-mode?

IIUC, [`treesit.el` is required in order to use `c-ts-mode`](https://github.com/emacs-mirror/emacs/blob/74ceb6922c621cf48a2c9f6e6622133b522252ab/lisp/progmodes/c-ts-mode.el#L69) -- there are other `*-ts-mode.el` files too for other languages.  I think these were all created after / while `treesit.el` was created.

I haven't used `tree-sitter.el` in a while -- perhaps someone else can answer regarding the "minor mode" part of your question.

**Reply by DrScKAWAMOTO** [reply]
Thanx Esq. sogaiu.

My emacs have TREE_SITTER feature.
```example
Configured features:
ACL GLIB GMP GNUTLS JPEG JSON LIBXML2 MODULES NOTIFY KQUEUE NS PDUMPER
PNG RSVG SQLITE3 THREADS TIFF TOOLKIT_SCROLL_BARS TREE_SITTER ZLIB
```
And in `*scratch*` buffer  I execute `(treesit-available-p)` then get result `t`.

I think that I am able to use treesit.el, but I have indentation trouble in c-ts-mode.
So I think that I would like to use tree-sitter.el rather than treesit.el.
May I use tree-sitter.el ?

**Reply by sogaiu** [reply]
As far as I understand, I think it's technically possible for you to use `tree-sitter.el`, provided your emacs settings are appropriate.  I'm not sure of the setting details though as I haven't been using elisp-tree-sitter recently (^^;

---

Regarding automatic indentation, I think with a matching closing brace, e.g.:

```c
void test(void)
{
  do
}               // matching closing brace
```

I get indentation that seems to match what you wanted when using `c-ts-mode` (so this is with `treesit.el`).

Without a closing brace, it doesn't seem to work so well here.

This could be because without "balanced" delimiters, the parse tree that tree-sitter comes up with is inappropriate for correct indentation -- the code without the matching closing brace is syntactically invalid afterall (though ofc, in the above case, just adding the brace doesn't make things valid either).  I suspect whether you use `treesit.el` or `tree-sitter.el` the results could be similar.

It may be that if you use something like `electric-pair-mode` the matching closing braces will get inserted for you automatically [1].  I know not everyone likes that kind of thing, but FWIW.

---

[1] Some related info [here](https://stackoverflow.com/questions/3801147/how-can-can-i-get-emacs-to-insert-closing-braces-automatically).

**Reply by DrScKAWAMOTO** [reply]
I added the closing parenthesis by hand and tried again.

From the next state
```c
void test(void)
{
  do🔳
}
```
If I write `Enter func();` then
```c
void test(void)
{
  do
  func();🔳
}
```
Then write 'Enter while (1);`
```c
void test(void)
{
  do
  func();
  while(1);🔳
}
```
That is not the indentation I intended.

**Reply by sogaiu** [reply]
It's been a while since I've used `do` in C, but isn't it necessary to have more curly braces, like this:

```c
void test(void)
{
  do {
    func();
  } while (1);
}
```

With the curly braces in place, does it work ok for you?

**Reply by DrScKAWAMOTO** [reply]
I am using gnu style (M-x c-ts-mode-set-style)
```c
void test(void)
{
  do
  {
    func();
  }
  while (1);
}
```
is the result.
I needed one is the following,
```c
void test(void)
{
  do
    {
      func();
    }
  while (1);
}
```

**Reply by DrScKAWAMOTO** [reply]
These operations are ok in c-mode.
Even in c-ts-mode, if I use tab to format it, it will work as I intended.

**Reply by DrScKAWAMOTO** [reply]
We have digressed into a topic that has nothing to do with tree-sitter.el.
I would like to try writing an init.el that uses tree-sitter.el.

Thanx Esq. sogaiu.
Thanx everyone.

**Reply by sogaiu** [reply]
Regarding `c-ts-mode`'s behavior, I suggest asking or reporting on one of the gnu emacs mailing lists, perhaps [bug-gnu-emacs](https://lists.gnu.org/archive/html/bug-gnu-emacs/) would be appropriate.

I see in some search results of the archives there that [`c-ts-mode` is being discussed](https://lists.gnu.org/archive/cgi-bin/namazu.cgi?query=c-ts-mode&submit=Search%21&idxname=bug-gnu-emacs&max=20&result=normal&sort=score).

---

I don't know if `tree-sitter.el` and friends work any better and I don't have a good way of testing that out ATM.  May be someone else knows though.
